File: zig/ambler.zig

const std = @import("std");

pub const Next = struct {
    run: *const fn () anyerror!?*const Next,
};

pub fn amble(initial: *const fn (i32) anyerror!?*const Next, state: i32) anyerror!void {
    var next = try initial(state);
    while (next) |n| {
        next = try n.run();
    }
}

File: go/pkg/ambler/ambler.go
package ambler

import (
	"github.com/argenkiwi/ambler/go/pkg/lead"
)

func Amble[S any](state S, next lead.Lead, follow func(lead.Lead, S) (S, lead.Lead, error)) (S, error) {
	currentState, nextLead, err := follow(next, state)
	if err != nil {
		return currentState, err
	}

	if nextLead == lead.None {
		return currentState, nil
	}

	return Amble(currentState, nextLead, follow)
}
